非*注解*配置的Reference代理服务，spring-boot的placeholder无法正常解析

dubbo version: 2.7.2
dubbo-spring-boot version: 2.7.1
spring-boot version 2.1.6.Release

启动类如下
```java

@EnableDubbo
@SpringBootApplication
@ImportResource("classpath:dubbo-config.xml")
public class Application {

    public static void main(String[] args) {
        SpringApplication.run(Application.class, args);
    }

}

```

Spring Boot配置文件

```yml
spring:
  application:
    name: first-dubbo-app-consumer
server:
  port: 8085

dubbo:
  protocol:
    name: dubbo
    port: 20880
  registry:
    address: zookeeper://${zookeeper.address:127.0.0.1:2181}
    session: 10000
  consumer:
    check: false
  metadata-report:
    address: zookeeper://${zookeeper.address:127.0.0.1:2181}
  config-center:
    address: zookeeper://${zookeeper.address:127.0.0.1:2181}

management:
  endpoints:
    web:
      exposure:
        include: "*"

```

Dubbo XML配置文件

```xml
<beans xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:dubbo="http://dubbo.apache.org/schema/dubbo"
    xmlns="http://www.springframework.org/schema/beans"
    xmlns:p="http://www.springframework.org/schema/p"
    xsi:schemaLocation="http://www.springframework.org/schema/beans
        http://www.springframework.org/schema/beans/spring-beans.xsd
        http://dubbo.apache.org/schema/dubbo
        http://dubbo.apache.org/schema/dubbo/dubbo.xsd">

    <dubbo:reference id="helloService" interface="org.chris.demo.dubbo.api.HelloService" generic="true" />

</beans>
```

因为一些原因，希望使用xml文件配置一些服务，但引入xml配置后，配置中心的地址被错误的解析了。报错信息如下。
```
Caused by: java.lang.NumberFormatException: For input string: "2181}"
	at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65) ~[na:1.8.0_201]
	at java.lang.Integer.parseInt(Integer.java:580) ~[na:1.8.0_201]
	at java.lang.Integer.parseInt(Integer.java:615) ~[na:1.8.0_201]
	at org.apache.dubbo.common.URL.valueOf(URL.java:269) ~[dubbo-2.7.2.jar:2.7.2]
	at org.apache.dubbo.common.utils.UrlUtils.parseURL(UrlUtils.java:100) ~[dubbo-2.7.2.jar:2.7.2]
	at org.apache.dubbo.config.ConfigCenterConfig.toUrl(ConfigCenterConfig.java:102) ~[dubbo-2.7.2.jar:2.7.2]
	at org.apache.dubbo.config.AbstractInterfaceConfig.prepareEnvironment(AbstractInterfaceConfig.java:289) ~[dubbo-2.7.2.jar:2.7.2]
	at org.apache.dubbo.config.AbstractInterfaceConfig.startConfigCenter(AbstractInterfaceConfig.java:279) ~[dubbo-2.7.2.jar:2.7.2]
	at org.apache.dubbo.config.ReferenceConfig.checkAndUpdateSubConfigs(ReferenceConfig.java:220) ~[dubbo-2.7.2.jar:2.7.2]
	at org.apache.dubbo.config.ReferenceConfig.get(ReferenceConfig.java:244) ~[dubbo-2.7.2.jar:2.7.2]
	at org.apache.dubbo.config.spring.beans.factory.annotation.ReferenceAnnotationBeanPostProcessor$ReferenceBeanInvocationHandler.init(ReferenceAnnotationBeanPostProcessor.java:173) ~[dubbo-2.7.2.jar:2.7.2]
	at org.apache.dubbo.config.spring.beans.factory.annotation.ReferenceAnnotationBeanPostProcessor$ReferenceBeanInvocationHandler.access$100(ReferenceAnnotationBeanPostProcessor.java:146) ~[dubbo-2.7.2.jar:2.7.2]
	at org.apache.dubbo.config.spring.beans.factory.annotation.ReferenceAnnotationBeanPostProcessor.buildInvocationHandler(ReferenceAnnotationBeanPostProcessor.java:140) ~[dubbo-2.7.2.jar:2.7.2]
	at org.apache.dubbo.config.spring.beans.factory.annotation.ReferenceAnnotationBeanPostProcessor.buildProxy(ReferenceAnnotationBeanPostProcessor.java:123) ~[dubbo-2.7.2.jar:2.7.2]
	at org.apache.dubbo.config.spring.beans.factory.annotation.ReferenceAnnotationBeanPostProcessor.doGetInjectedBean(ReferenceAnnotationBeanPostProcessor.java:119) ~[dubbo-2.7.2.jar:2.7.2]
	at org.apache.dubbo.config.spring.beans.factory.annotation.ReferenceAnnotationBeanPostProcessor.doGetInjectedBean(ReferenceAnnotationBeanPostProcessor.java:52) ~[dubbo-2.7.2.jar:2.7.2]
	at org.apache.dubbo.config.spring.beans.factory.annotation.AnnotationInjectedBeanPostProcessor.getInjectedObject(AnnotationInjectedBeanPostProcessor.java:337) ~[dubbo-2.7.2.jar:2.7.2]
	at org.apache.dubbo.config.spring.beans.factory.annotation.AnnotationInjectedBeanPostProcessor$AnnotatedFieldElement.inject(AnnotationInjectedBeanPostProcessor.java:517) ~[dubbo-2.7.2.jar:2.7.2]
	at org.springframework.beans.factory.annotation.InjectionMetadata.inject(InjectionMetadata.java:90) ~[spring-beans-5.1.8.RELEASE.jar:5.1.8.RELEASE]
	at org.apache.dubbo.config.spring.beans.factory.annotation.AnnotationInjectedBeanPostProcessor.postProcessPropertyValues(AnnotationInjectedBeanPostProcessor.java:130) ~[dubbo-2.7.2.jar:2.7.2]
	... 17 common frames omitted
```

如果将xml配置去掉，dubbo就能够正确解析spring-boot配置文件中的内容。是不是xml文件加载的配置信息早于spring-boot环境启动，造成spring-boot的环境配置没有完全加载完。

--- a/dubbo-spring-boot-actuator/src/test/java/org/apache/dubbo/spring/boot/actuate/autoconfigure/DubboEndpointAnnotationAutoConfigurationTest.java
+++ b/dubbo-spring-boot-actuator/src/test/java/org/apache/dubbo/spring/boot/actuate/autoconfigure/DubboEndpointAnnotationAutoConfigurationTest.java
@@ -160,7 +160,7 @@ public class DubboEndpointAnnotationAutoConfigurationTest {
 
         Assert.assertEquals(1, services.size());
 
-        Map<String, Object> demoServiceMeta = services.get("providers:my-protocol:org.apache.dubbo.spring.boot.actuate.autoconfigure.DubboEndpointAnnotationAutoConfigurationTest$DemoService:1.0.0");
+        Map<String, Object> demoServiceMeta = services.get("ServiceBean:org.apache.dubbo.spring.boot.actuate.autoconfigure.DubboEndpointAnnotationAutoConfigurationTest$DemoService:1.0.0");
 
         Assert.assertEquals("1.0.0", demoServiceMeta.get("version"));
 

--- a/dubbo-spring-boot-compatible/actuator/src/test/java/org/apache/dubbo/spring/boot/actuate/autoconfigure/DubboEndpointAutoConfigurationTest.java
+++ b/dubbo-spring-boot-compatible/actuator/src/test/java/org/apache/dubbo/spring/boot/actuate/autoconfigure/DubboEndpointAutoConfigurationTest.java
@@ -164,7 +164,7 @@ public class DubboEndpointAutoConfigurationTest {
 
         Assert.assertEquals(1, services.size());
 
-        Map<String, Object> demoServiceMeta = services.get("providers:my-protocol:org.apache.dubbo.spring.boot.actuate.autoconfigure.DubboEndpointAutoConfigurationTest$DemoService:1.0.0");
+        Map<String, Object> demoServiceMeta = services.get("ServiceBean:org.apache.dubbo.spring.boot.actuate.autoconfigure.DubboEndpointAutoConfigurationTest$DemoService:1.0.0");
 
         Assert.assertEquals("1.0.0", demoServiceMeta.get("version"));
 

--- a/dubbo-spring-boot-compatible/autoconfigure/src/main/java/org/apache/dubbo/spring/boot/util/DubboUtils.java
+++ b/dubbo-spring-boot-compatible/autoconfigure/src/main/java/org/apache/dubbo/spring/boot/util/DubboUtils.java
@@ -23,7 +23,6 @@ import org.apache.dubbo.config.spring.context.annotation.EnableDubboConfigBindin
 import org.apache.dubbo.config.spring.context.properties.DubboConfigBinder;
 
 import org.springframework.boot.context.ContextIdApplicationContextInitializer;
-import org.springframework.core.env.AbstractEnvironment;
 import org.springframework.core.env.ConfigurableEnvironment;
 import org.springframework.core.env.PropertyResolver;
 
@@ -31,7 +30,6 @@ import java.util.Collections;
 import java.util.Map;
 import java.util.SortedMap;
 import java.util.TreeMap;
-import java.util.stream.Collectors;
 
 /**
  * The utilities class for Dubbo
@@ -191,29 +189,12 @@ public abstract class DubboUtils {
 
             if (propertyName.startsWith(DUBBO_PREFIX + PROPERTY_NAME_SEPARATOR)
                     && entry.getValue() != null) {
-                dubboProperties.put(propertyName, entry.getValue().toString());
+                dubboProperties.put(propertyName, environment.resolvePlaceholders(entry.getValue().toString()));
             }
 
         }
 
-        return Collections.unmodifiableSortedMap(dubboProperties.entrySet().stream()
-                .collect(Collectors.toMap(
-                        Map.Entry::getKey,
-                        DubboUtils::resolvePlaceholders,
-                        (u,v) -> { throw new IllegalStateException(String.format("Duplicate key %s", u)); } ,
-                        () -> new TreeMap<>())));
-
+        return Collections.unmodifiableSortedMap(dubboProperties);
     }
 
-    private static Object resolvePlaceholders(Map.Entry<String, Object> entry) {
-        if (entry == null) {
-            return null;
-        }
-        Object value = entry.getValue();
-        if (value != null && value instanceof String) {
-            AbstractEnvironment resolver = new AbstractEnvironment() {};
-            return resolver.resolvePlaceholders((String) value);
-        }
-        return value;
-    }
 }

--- a/dubbo-spring-boot-compatible/autoconfigure/src/main/java/org/apache/dubbo/spring/boot/util/DubboUtils.java
+++ b/dubbo-spring-boot-compatible/autoconfigure/src/main/java/org/apache/dubbo/spring/boot/util/DubboUtils.java
@@ -23,6 +23,7 @@ import org.apache.dubbo.config.spring.context.annotation.EnableDubboConfigBindin
 import org.apache.dubbo.config.spring.context.properties.DubboConfigBinder;
 
 import org.springframework.boot.context.ContextIdApplicationContextInitializer;
+import org.springframework.core.env.AbstractEnvironment;
 import org.springframework.core.env.ConfigurableEnvironment;
 import org.springframework.core.env.PropertyResolver;
 
@@ -30,6 +31,7 @@ import java.util.Collections;
 import java.util.Map;
 import java.util.SortedMap;
 import java.util.TreeMap;
+import java.util.stream.Collectors;
 
 /**
  * The utilities class for Dubbo
@@ -194,8 +196,24 @@ public abstract class DubboUtils {
 
         }
 
-        return Collections.unmodifiableSortedMap(dubboProperties);
+        return Collections.unmodifiableSortedMap(dubboProperties.entrySet().stream()
+                .collect(Collectors.toMap(
+                        Map.Entry::getKey,
+                        DubboUtils::resolvePlaceholders,
+                        (u,v) -> { throw new IllegalStateException(String.format("Duplicate key %s", u)); } ,
+                        () -> new TreeMap<>())));
 
     }
 
+    private static Object resolvePlaceholders(Map.Entry<String, Object> entry) {
+        if (entry == null) {
+            return null;
+        }
+        Object value = entry.getValue();
+        if (value != null && value instanceof String) {
+            AbstractEnvironment resolver = new AbstractEnvironment() {};
+            return resolver.resolvePlaceholders((String) value);
+        }
+        return value;
+    }
 }

